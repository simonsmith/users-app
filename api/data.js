const faker = require('faker');

module.exports = () => {
  return {
    users: Array.from({length: 10}, () => faker.helpers.createCard()),
  };
};
