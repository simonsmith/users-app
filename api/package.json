{
  "name": "api",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "scripts": {
    "start": "json-server --port 5000 data.js"
  },
  "dependencies": {
    "faker": "^4.1.0",
    "json-server": "^0.15.0",
    "lodash": "^4.17.11",
    "nanoid": "^2.0.3"
  },
  "devDependencies": {}
}
