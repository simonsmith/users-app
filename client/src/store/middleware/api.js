import axios from 'axios';
import {isFunction} from 'lodash-fp';

export default function apiMiddleware({dispatch}) {
  return next => action => {
    if (!action['API_REQUEST']) {
      return next(action);
    }
    const {
      type,
      meta: {url, method, transformer},
    } = action['API_REQUEST'];
    next({type: `${type}_REQUEST`});
    return axios({
      method,
      url,
    })
      .then(
        response => {
          const data = isFunction(transformer)
            ? transformer(response.data)
            : response.data;
          dispatch({
            type: `${type}_${method.toUpperCase()}_SUCCESS`,
            payload: data,
          });
        },
        error => {
          dispatch({type: `${type}_FAILURE`, error: true, payload: error});
        }
      )
      .catch(e => console.error(e));
  };
}

export function makeApiRequest({
  method = 'get',
  url,
  type,
  transformer = x => x,
}) {
  return {
    API_REQUEST: {
      type,
      meta: {method, url, transformer},
    },
  };
}
