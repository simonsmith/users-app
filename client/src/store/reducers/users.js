import {createReducer} from 'redux-starter-kit';

const initialState = {
  isLoading: false,
  result: [],
  error: null,
};

export default createReducer(initialState, {
  USERS_REQUEST: state => {
    state.isLoading = true;
  },
  USERS_GET_SUCCESS: (state, action) => {
    state.isLoading = false;
    state.result = action.payload.result;
  },
  USERS_FAILURE: (state, action) => {
    state.isLoading = false;
    if (action.error) {
      state.error = action.payload;
    }
  },
});
