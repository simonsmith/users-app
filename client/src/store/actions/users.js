import {makeApiRequest} from '../middleware/api';

const URL = 'http://localhost:5000/users';

export function requestAllUsers() {
  return makeApiRequest({
    url: URL,
    type: 'USERS',
  });
}
