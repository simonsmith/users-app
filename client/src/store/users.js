import {createReducer} from 'redux-starter-kit';
import {normalize, schema} from 'normalizr';
import {apiRequest} from './middleware/api';

export const FETCH_USERS = 'FETCH_USERS';
export const CREATE_USER = 'CREATE_USER';
export const FETCH_USERS_SUCCESS = 'FETCH_USERS_SUCCESS';

const role = new schema.Entity('role');
const user = new schema.Entity('users', {
  role,
});

function normalizeUsers({data}) {
  return {
    type: FETCH_USERS_SUCCESS,
    payload: normalize(data, [user]),
  };
}

export function createNewUser(user) {
  return apiRequest('/users', {
    type: CREATE_USER,
    onSuccess: requestAllUsers,
    method: 'POST',
    data: user,
  });
}

export function requestAllUsers() {
  return apiRequest('/users', {
    type: FETCH_USERS,
    onSuccess: normalizeUsers,
  });
}

const initialState = {
  result: [],
};

export const usersReducer = createReducer(initialState, {
  [FETCH_USERS_SUCCESS]: (state, action) => {
    state.result = action.payload.result;
  },
});

export function selectUsers(state) {
  return {
    result: state.users.result,
    entities: state.entities.users,
  };
}

export function selectRoles(state) {
  return {
    entities: state.entities.role,
  };
}
