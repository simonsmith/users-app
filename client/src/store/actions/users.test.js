import {requestAllUsers} from './users';
import {makeApiRequest} from '../middleware/api';

jest.mock('../middleware/api');

test('calls makeApiRequest with the correct arguments', () => {
  requestAllUsers();
  expect(makeApiRequest).toMatchSnapshot();
});

test('transforms data with normalizr', () => {
  const data = [
    {id: '123', name: 'foo', role: {id: '999', title: 'hr'}},
    {id: '456', name: 'bar', role: {id: '888', title: 'hr'}},
    {id: '789', name: 'baz', role: {id: '777', title: 'cleaner'}},
  ];
  requestAllUsers();
  const {transformer} = makeApiRequest.mock.calls[0][0];
  expect(transformer(data)).toMatchSnapshot();
});
