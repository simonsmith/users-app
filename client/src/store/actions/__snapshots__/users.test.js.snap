// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls makeApiRequest with the correct arguments 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "transformer": [Function],
        "type": "USERS",
        "url": "http://localhost:5000/users",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`transforms data with normalizr 1`] = `
Object {
  "entities": Object {
    "role": Object {
      "777": Object {
        "id": "777",
        "title": "cleaner",
      },
      "888": Object {
        "id": "888",
        "title": "hr",
      },
      "999": Object {
        "id": "999",
        "title": "hr",
      },
    },
    "users": Object {
      "123": Object {
        "id": "123",
        "name": "foo",
        "role": "999",
      },
      "456": Object {
        "id": "456",
        "name": "bar",
        "role": "888",
      },
      "789": Object {
        "id": "789",
        "name": "baz",
        "role": "777",
      },
    },
  },
  "result": Array [
    "123",
    "456",
    "789",
  ],
}
`;
